File: 0421.MaximumXORofTwoNumbersinanArray.py
class Solution(object):
    def findMaximumXOR(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ans = 0
        mask = 0
        for i in xrange(31, -1, -1):
            mask |= (1 << i)
            Set = {num & mask for num in nums}
            '''
            Set = set()
            for num in nums:
                Set.add(num & mask)
            '''
            tmp = ans | (1 << i)
            for prefix in Set:
                if tmp ^ prefix in Set:
                    ans = tmp
                    break
        return ans
        

File: 0421.MaximumXORofTwoNumbersinanArray.go
func findMaximumXOR(nums []int) int {
    ans, mask := 0, 0
    for i := 31; i >= 0; i-- {
        mask |= 1 << uint(i)
        Map := make(map[int]bool)
        for _, num := range nums {
            Map[num&mask] = true
        }
        tmp := ans | 1<<uint(i)
        for prefix := range Map {
            if Map[prefix^tmp] {
                ans = tmp
                break
            }
        }
    }
    return ans
}

func findMaximumAND(nums []int) int {
    ans := 0
    for bit := 1<<31; bit >= 1; bit >>= 1 {
        nxt := []int{}
        for _, num := range nums {
            if num & bit > 0 {
                nxt = append(nxt, num)
            }
        }
        if len(nxt) >= 2 {
            ans |= bit
            nums = nxt
        }
    }
    return ans
}
